taichi/ui/backends/vulkan/vulkan_utils.cpp

```diff
@@ -0,0 +1,472 @@
+
+  alloc_device_memory(mem_requirements, properties, device, physical_device,
+                      buffer_mem);
+
+  vkBindBufferMemory(device, buffer, buffer_mem, 0);
+}
+
+VkShaderModule create_shader_module(const std::vector<char> &code,
+                                    VkDevice device) {
+  VkShaderModuleCreateInfo create_info{};
+  create_info.sType = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO;
+  create_info.codeSize = code.size();
+  create_info.pCode = reinterpret_cast<const uint32_t *>(code.data());
+
+  VkShaderModule shader_module;
+  if (vkCreateShaderModule(device, &create_info, nullptr, &shader_module) !=
+      VK_SUCCESS) {
+    throw std::runtime_error("failed to create shader module!");
+  }
+
+  return shader_module;
+}
+
+void create_image(int dim,
+                  uint32_t width,
+                  uint32_t height,
+                  uint32_t depth,
+                  VkFormat format,
+                  VkImageTiling tiling,
+                  VkImageUsageFlags usage,
+                  VkMemoryPropertyFlags properties,
+                  VkImage &image,
+                  VkDeviceMemory &image_mem,
+                  VkDevice device,
+                  VkPhysicalDevice physical_device) {
+  VkImageCreateInfo image_info{};
+  image_info.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
+  if (dim == 1) {
+    image_info.imageType = VK_IMAGE_TYPE_1D;
+  } else if (dim == 2) {
+    image_info.imageType = VK_IMAGE_TYPE_2D;
+  } else if (dim == 3) {
+    image_info.imageType = VK_IMAGE_TYPE_3D;
+  } else {
+    throw std::runtime_error("dim can only be 1 2 3");
+  }
+  image_info.extent.width = width;
+  image_info.extent.height = height;
+  image_info.extent.depth = depth;
+  image_info.mipLevels = 1;
+  image_info.arrayLayers = 1;
+  image_info.format = format;
+  image_info.tiling = tiling;
+  image_info.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+  image_info.usage = usage;
+  image_info.samples = VK_SAMPLE_COUNT_1_BIT;
+  image_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+
+  VkExternalMemoryImageCreateInfo external_mem_image_create_info = {};
+  external_mem_image_create_info.sType =
+      VK_STRUCTURE_TYPE_EXTERNAL_MEMORY_IMAGE_CREATE_INFO;
+  external_mem_image_create_info.pNext = NULL;
+
+#ifdef _WIN64
+  external_mem_image_create_info.handleTypes =
+      VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_BIT;
+#else
+  external_mem_image_create_info.handleTypes =
+      VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD_BIT_KHR;
+#endif
+
+  image_info.pNext = &external_mem_image_create_info;
+
+  if (vkCreateImage(device, &image_info, nullptr, &image) != VK_SUCCESS) {
+    throw std::runtime_error("failed to create image!");
+  }
+
+  VkMemoryRequirements mem_requirements;
+  vkGetImageMemoryRequirements(device, image, &mem_requirements);
+
+  alloc_device_memory(mem_requirements, properties, device, physical_device,
+                      image_mem);
+
+  vkBindImageMemory(device, image, image_mem, 0);
+}
```

Comment:
nit: use Vulkan memory allocator from the vulkan backend, and move to our Device API in the future. (This & buffer)

Reply:
I could use a bit of help regarding this. For CUDA-Vk interop, we need the corresponding `VkDeviceMemory` object and the offset for each buffer and image. The current memory allocator does not appear to expose these. Could you help add these to the allocator? As I'm not very familiar with vma.